### PR TITLE
[PPDiffuers] Add CycleDiffusion based on FastDeploy

### PR DESCRIPTION
**Pull request by joey12300**
### PR types

New features
### PR changes

Models
### Description

Add CycleDiffusion based on FastDeploy. 

### Benchmark
Run CycleDiffusionPipeline 10 times and get average latency to compare the performace between  FastDeploy version and pytorch version.

|Args| value|
|---|---|
|batch size | 1 |
| num_inference_steps | 100 |
| strength | 0.8 |
| eta | 0.1 |
| guidance_scale | 2 |
| source_guidance_scale | 1 |
| source_prompt | “An astronaut riding a horse” |
| prompt | "An astronaut riding an elephant" |

#### Average latency
|FastDeploy + PaddleTRT FP16| Diffusers + Torch FP16|
|---|---|
|6.66s|12.86s|

### CODE REVIEW

File: ppdiffusers/ppdiffusers/pipelines/stable_diffusion/pipeline_fastdeploy_cycle_diffusion.py [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+                padding="max_length",
+                max_length=self.tokenizer.model_max_length,
+                truncation=True,
+                return_tensors="np",
+            )
+            text_input_ids = text_inputs.input_ids
+            untruncated_ids = self.tokenizer(prompt, padding="longest", return_tensors="np").input_ids
+
+            if untruncated_ids.shape[-1] >= text_input_ids.shape[-1] and not paddle.equal_all(
+                text_input_ids, untruncated_ids
```
Comment (guoshengCS) [inline ppdiffusers/ppdiffusers/pipelines/stable_diffusion/pipeline_fastdeploy_cycle_diffusion.py]: `text_input_ids`和`untruncated_ids`是np的，这里paddle.equal_all是否需要用np的呢

File: ppdiffusers/ppdiffusers/pipelines/stable_diffusion/pipeline_fastdeploy_cycle_diffusion.py [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+                )
+                source_latents = prev_source_latents
+
+                # compute the previous noisy sample x_t -> x_t-1
+                latents = self.scheduler.step(
+                    noise_pred, t, latents, variance_noise=noise, **extra_step_kwargs
+                ).prev_sample
+                if i == len(timesteps) - 1:
+                    # sync for accuracy it/s measure
+                    paddle.device.cuda.synchronize()
```
Comment (guoshengCS) [inline ppdiffusers/ppdiffusers/pipelines/stable_diffusion/pipeline_fastdeploy_cycle_diffusion.py:675]: 这里是否是必须的呢，看其他的pipelile包括FastDeploy pipeline里没有这个

**Review by guoshengCS — DISMISSED** [review]

File: ppdiffusers/ppdiffusers/pipelines/stable_diffusion/pipeline_fastdeploy_cycle_diffusion.py (hunk 2 above)
Comment (joey12300) [reply ppdiffusers/ppdiffusers/pipelines/stable_diffusion/pipeline_fastdeploy_cycle_diffusion.py:675]: 主要保证多流kernel异步发射的最终结果正确，需要同步

File: ppdiffusers/ppdiffusers/pipelines/stable_diffusion/pipeline_fastdeploy_cycle_diffusion.py (hunk 1 above)
Comment (joey12300) [reply ppdiffusers/ppdiffusers/pipelines/stable_diffusion/pipeline_fastdeploy_cycle_diffusion.py]: 感谢review，已经更新了~

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+        option.set_external_raw_stream(paddle_stream)
+    for pass_name in disable_paddle_pass:
+        option.paddle_infer_option.delete_pass(pass_name)
+    if use_trt:
+        option.paddle_infer_option.disable_trt_ops(disable_paddle_trt_ops)
+        option.paddle_infer_option.enable_trt = True
+        if use_fp16:
+            option.trt_option.enable_fp16 = True
+        cache_file = os.path.join(model_dir, model_prefix, "inference.trt")
+        option.set_trt_cache_file(cache_file)
```
Comment (jiangjiajun) [inline ppdiffusers/deploy/text_guided_img_to_img_infer.py]: 看到上面的代码已经开始在使用1.0.4的API了， `option.paddle_infer_option`，那就统一都开始切为新版本API吧

改为`option.trt_option.serialize_file = cache_file`

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+        # Need to enable collect shape for ernie
+        if dynamic_shape is not None:
+            option.enable_paddle_trt_collect_shape()
```
Comment (jiangjiajun) [inline ppdiffusers/deploy/text_guided_img_to_img_infer.py]: `option.paddle_infer_option.collect_trt_shape = True`

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+            for key, shape_dict in dynamic_shape.items():
+                option.set_trt_input_shape(
+                    key,
```
Comment (jiangjiajun) [inline ppdiffusers/deploy/text_guided_img_to_img_infer.py]: option.trt_option.set_shape(name, min, opt, max)

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+    option.set_model_path(model_file, params_file)
+    return fd.Runtime(option)
+
+
+def create_paddle_lite_runtime(model_dir, model_prefix, device="cpu", device_id=0):
+    option = fd.RuntimeOption()
+    option.use_lite_backend()
+    if device == "huawei_ascend_npu":
+        option.use_ascend()
+        option.set_lite_device_names(["huawei_ascend_npu"])
```
Comment (jiangjiajun) [inline ppdiffusers/deploy/text_guided_img_to_img_infer.py]: 此行代码应该无需显式调用，删除即可

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+        option.set_lite_model_cache_dir(os.path.join(model_dir, model_prefix))
+        option.set_lite_context_properties(
+            "HUAWEI_ASCEND_NPU_SELECTED_DEVICE_IDS={};HUAWEI_ASCEND_NPU_PRECISION_MODE=allow_mix_precision".format(
```
Comment (jiangjiajun) [inline ppdiffusers/deploy/text_guided_img_to_img_infer.py:147]:
```
option.paddle_lite_option.nnadapter_model_cache_dir = os.path.join(model_dir, model_prefix)
option.paddle_lite_option.nnadapter_context_properties = "HUAWEI_ASCEND_NPU_SELECTED_DEVICE_IDS={};HUAWEI_ASCEND_NPU_PRECISION_MODE=allow_mix_precision".format(device_id)
```

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+    params_file = os.path.join(model_dir, model_prefix, "inference.pdiparams")
+    option.set_model_path(model_file, params_file)
+    return fd.Runtime(option)
+
+
+def create_trt_runtime(model_dir, model_prefix, model_format, workspace=(1 << 31), dynamic_shape=None, device_id=0):
+    option = fd.RuntimeOption()
+    option.use_trt_backend()
+    option.use_gpu(device_id)
+    option.enable_trt_fp16()
```
Comment (jiangjiajun) [inline ppdiffusers/deploy/text_guided_img_to_img_infer.py]: option.trt_option.enable_fp16 = True

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+    option.set_trt_max_workspace_size(workspace)
```
Comment (jiangjiajun) [inline ppdiffusers/deploy/text_guided_img_to_img_infer.py]: option.trt_option.max_workspace_size = workspace

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+    if dynamic_shape is not None:
+        for key, shape_dict in dynamic_shape.items():
+            option.set_trt_input_shape(
```
Comment (jiangjiajun) [inline ppdiffusers/deploy/text_guided_img_to_img_infer.py]: option.set_shape

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+            )
+    if model_format == "paddle":
+        model_file = os.path.join(model_dir, model_prefix, "inference.pdmodel")
+        params_file = os.path.join(model_dir, model_prefix, "inference.pdiparams")
+        option.set_model_path(model_file, params_file)
+    else:
+        onnx_file = os.path.join(model_dir, model_prefix, "inference.onnx")
+        option.set_model_path(onnx_file, model_format=ModelFormat.ONNX)
+    cache_file = os.path.join(model_dir, model_prefix, "inference.trt")
+    option.set_trt_cache_file(cache_file)
```
Comment (jiangjiajun) [inline ppdiffusers/deploy/text_guided_img_to_img_infer.py]: option.trt_option.serialize_file = cache_file

**Review by jiangjiajun — CHANGES_REQUESTED** [review]

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py (hunk 3 above)
Comment (joey12300) [reply ppdiffusers/deploy/text_guided_img_to_img_infer.py]: Done

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py (hunk 5 above)
Comment (joey12300) [reply ppdiffusers/deploy/text_guided_img_to_img_infer.py]: done

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py (hunk 6 above)
Comment (joey12300) [reply ppdiffusers/deploy/text_guided_img_to_img_infer.py]: Done

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py (hunk 4 above)
Comment (joey12300) [reply ppdiffusers/deploy/text_guided_img_to_img_infer.py]: Done

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py (hunk 7 above)
Comment (joey12300) [reply ppdiffusers/deploy/text_guided_img_to_img_infer.py:147]: Done

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py (hunk 11 above)
Comment (joey12300) [reply ppdiffusers/deploy/text_guided_img_to_img_infer.py]: Done

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py (hunk 10 above)
Comment (joey12300) [reply ppdiffusers/deploy/text_guided_img_to_img_infer.py]: Done

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py (hunk 9 above)
Comment (joey12300) [reply ppdiffusers/deploy/text_guided_img_to_img_infer.py]: Done

File: ppdiffusers/deploy/text_guided_img_to_img_infer.py (hunk 8 above)
Comment (joey12300) [reply ppdiffusers/deploy/text_guided_img_to_img_infer.py]: Done

**Review by guoshengCS — APPROVED** [review]

**Pull request merged**